### DISCUSSION TITLE
Large Integers Corrupted by the JSONRenderer

### DISCUSSION THREAD

**Discussion by Corban-Lee** [encode/django-rest-framework#9547] [Potential Issue] [answered]
I store Discord server ID's as primary keys for an object in my database, which is either Sqlite or Postgresql depending on my environment being production or development.

Storing and retrieving these integers is trivial and works without issue, including when fetching them through the DRF API web interface, however the JSON interface for retrieving data corrupts these integers.

Visual Representation (the primary key field is "id"):

![correct image with api view](https://github.com/user-attachments/assets/4a3683a6-b988-4e3e-ae58-e685b690cce6)
![incorrect image with json view](https://github.com/user-attachments/assets/7e8b188d-17c7-44f1-a0e1-5812fe2afdae)

In the second image, the primary key ("id") has been corrupted, so that the last 2 digits are rounded down to `00`.

I was able to overcome this issue with a custom JSON renderer, that casts large integers to strings, however this issue should probably be addressed within DRF itself.

~~As for the root of the issue, I strongly believe it's related to the json standard library in python, although I haven't looked further to confirm this.~~

**Comment by TGoddessana** [comment]
Can you tell me how I can reproduce the issue?

**Reply by Corban-Lee** [reply]
Ah, excuse me for forgetting to include reproducible examples!

I have a model with a `PositiveIntegerField` field, but this will also replicate with an `IntegerField`, `BigIntegerField`, `PositiveBigIntegerField`.

Here is a minimal reproducible code example:
```
# model

class Server(models.Model):
    id = models.PositiveIntegerField(primary_key=True)


# serializer

class ServerSerializer(serializers.ModelSerializer):
    class Meta:
        model = Server
        fields = ("id",)


# view

class ServerView(generics.RetrieveAPIView):
    serializer_class = ServerSerializer
    queryset = Server.objects.all()
    
  
# route

urlpatterns = [
    path("servers/", ServerView.as_view())
]
```

Now reproduce with:
1. Create an instance of `Server`, and store a large interger (such as `1204426362794811453`).
2. View the newly created server through the API route (in this example `/servers/`)
3. Check the differences of the `id` field between the API view and JSON view (querystring `?format=api` and `?format=json`)

The issue has been reproduced if you see a difference between the ids, one being as specified, the other usually replaced the last 2 characters with `00`.

**Comment by lovelydinosaur** [comment]
> As for the root of the issue, I strongly believe it's related to the json standard library in python, although I haven't looked further to confirm this.

I would suggest looking further to confirm this.

**Reply by Corban-Lee** [reply]
Excuse me, I made this assumption on bad data.
I have no reason to believe it relates to the built-in json library.

**Reply by lovelydinosaur** [reply]
> I was able to overcome this issue with a custom JSON renderer

Eg?...

**Comment by aayushostwal** [comment]
I just tried to replicate this issue, just made slight changes. 
I had to change the url pattern and added pk as param.
```python
path("server/<int:pk>/", ServerView.as_view()),
```

But after this changes, I have tried for `ID= 1204426362794811453` which worked completely fine.

I would require more insights on how you have coded this app.

**Reply by aayushostwal** [reply, marked as answer]
![image](https://github.com/user-attachments/assets/48f55821-d079-4a32-babb-c9d580ea354a)

![image](https://github.com/user-attachments/assets/42a63808-6544-4a01-8a2d-a8dd1c9548bf)
